File: app/store.py
"""
A tiny JSON-file "database" (data/log.json) for deduplication and history.
Committed back to the repo by GitHub Actions, so it costs nothing and lets the
app remember what it has already covered.
"""

from __future__ import annotations

import json
import re
from datetime import date, timedelta
from pathlib import Path


def fingerprint(headline: str) -> str:
    """A normalized fingerprint of a headline, used to detect repeat stories."""
    words = re.sub(r"[^a-z0-9]+", " ", headline.lower()).strip().split()
    return " ".join(words[:10])


def _day(offset_days: int = 0) -> str:
    return (date.today() + timedelta(days=offset_days)).isoformat()


def load_log(file: str | Path) -> dict:
    try:
        raw = json.loads(Path(file).read_text())
        entries = raw.get("entries", [])
        return {"entries": entries if isinstance(entries, list) else []}
    except (OSError, json.JSONDecodeError):
        return {"entries": []}


def save_log(file: str | Path, store: dict) -> None:
    """Save the log, trimmed to the last 30 days (and 400 entries) to stay small."""
    cutoff = _day(-30)
    entries = [e for e in store["entries"] if e.get("date", "") >= cutoff][-400:]
    Path(file).write_text(json.dumps({"entries": entries}, indent=2) + "\n")


def seen_recently(store: dict, headline: str, days: int = 2) -> bool:
    """True if a story with this headline was already logged in the last `days`."""
    fp = fingerprint(headline)
    cutoff = _day(-(days - 1))
    return any(
        e.get("fp") == fp and e.get("date", "") >= cutoff for e in store["entries"]
    )


def log_story(store: dict, entry: dict) -> None:
    """Add a story to the log (skipped if the same story is already logged today)."""
    entry_date = entry.get("date") or _day(0)
    fp = fingerprint(entry["headline"])
    if any(
        e.get("fp") == fp and e.get("date") == entry_date for e in store["entries"]
    ):
        return
    store["entries"].append({**entry, "fp": fp, "date": entry_date})


def medium_last_week(store: dict) -> list[dict]:
    """Unique MEDIUM-severity stories logged in the last 7 days — the weekly roundup."""
    cutoff = _day(-6)
    seen: set[str] = set()
    out: list[dict] = []
    for e in store["entries"]:
        if (
            e.get("date", "") < cutoff
            or e.get("severity") != "MEDIUM"
            or e.get("fp") in seen
        ):
            continue
        seen.add(e["fp"])
        out.append(e)
    return out

File: src/lib/store.ts
// A tiny JSON-file "database" (data/log.json) for deduplication and history.
// Committed back to the repo by GitHub Actions, so it costs nothing and lets
// the app remember what it has already covered.

import { promises as fs } from "fs";
import type { Severity } from "./types";

export interface LogEntry {
  fp: string; // fingerprint of the headline
  headline: string;
  summary: string;
  why: string;
  domain: string;
  severity: Severity;
  date: string; // YYYY-MM-DD
  mode: string; // morning | afternoon | evening | monitor
}

export interface LogStore {
  entries: LogEntry[];
}

/** A normalized fingerprint of a headline, used to detect repeat stories. */
export function fingerprint(headline: string): string {
  return headline
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, " ")
    .trim()
    .split(" ")
    .filter(Boolean)
    .slice(0, 10)
    .join(" ");
}

function dayString(offsetDays = 0): string {
  const d = new Date();
  d.setUTCDate(d.getUTCDate() + offsetDays);
  return d.toISOString().slice(0, 10);
}

export async function loadLog(file: string): Promise<LogStore> {
  try {
    const raw = JSON.parse(await fs.readFile(file, "utf8")) as LogStore;
    return { entries: Array.isArray(raw.entries) ? raw.entries : [] };
  } catch {
    return { entries: [] };
  }
}

/** Save the log, trimmed to the last 30 days (and 400 entries) to stay small. */
export async function saveLog(file: string, store: LogStore): Promise<void> {
  const cutoff = dayString(-30);
  const entries = store.entries
    .filter((e) => e.date >= cutoff)
    .slice(-400);
  await fs.writeFile(file, JSON.stringify({ entries }, null, 2) + "\n");
}

/** True if a story with this headline was already logged in the last `days`. */
export function seenRecently(
  store: LogStore,
  headline: string,
  days = 2,
): boolean {
  const fp = fingerprint(headline);
  const cutoff = dayString(-(days - 1));
  return store.entries.some((e) => e.fp === fp && e.date >= cutoff);
}

/** Add a story to the log (skipped if the same story is already logged today). */
export function logStory(
  store: LogStore,
  entry: Omit<LogEntry, "fp" | "date"> & { date?: string },
): void {
  const date = entry.date || dayString(0);
  const fp = fingerprint(entry.headline);
  if (store.entries.some((e) => e.fp === fp && e.date === date)) return;
  store.entries.push({ ...entry, fp, date });
}

/** Unique MEDIUM-severity stories logged in the last 7 days — the weekly roundup. */
export function mediumLastWeek(store: LogStore): LogEntry[] {
  const cutoff = dayString(-6);
  const seen = new Set<string>();
  const out: LogEntry[] = [];
  for (const e of store.entries) {
    if (e.date < cutoff || e.severity !== "MEDIUM" || seen.has(e.fp)) continue;
    seen.add(e.fp);
    out.push(e);
  }
  return out;
}
